src/main/java/io/confluent/kafkarest/extension/KafkaRestContextProvider.java

```diff
@@ -0,0 +1,128 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.kafkarest.extension;
+
+import org.apache.kafka.common.security.JaasUtils;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import io.confluent.kafkarest.ConsumerManager;
+import io.confluent.kafkarest.KafkaRestConfig;
+import io.confluent.kafkarest.KafkaRestContext;
+import io.confluent.kafkarest.KafkaRestContextImpl;
+import io.confluent.kafkarest.MetadataObserver;
+import io.confluent.kafkarest.ProducerPool;
+import io.confluent.kafkarest.SimpleConsumerFactory;
+import io.confluent.kafkarest.SimpleConsumerManager;
+import io.confluent.kafkarest.v2.KafkaConsumerManager;
+import kafka.utils.ZkUtils;
+
+public class KafkaRestContextProvider {
+
+  private static KafkaRestContext context = null;
+  private static ZkUtils defaultZkUtils = null;
+  private static KafkaRestConfig defaultAppConfig = null;
+  private static final InheritableThreadLocal<KafkaRestContext> restContextInheritableThreadLocal =
+      new InheritableThreadLocal<>();
+
+  private static AtomicBoolean initialized = new AtomicBoolean();
+
+  public static void initialize(
+      ZkUtils zkUtils,
+      KafkaRestConfig appConfig,
+      MetadataObserver mdObserver,
+      ProducerPool producerPool,
+      ConsumerManager consumerManager,
+      SimpleConsumerFactory simpleConsumerFactory,
+      SimpleConsumerManager simpleConsumerManager,
+      KafkaConsumerManager kafkaConsumerManager
+  ) {
+    if (initialized.compareAndSet(false, true)) {
+      if (zkUtils == null) {
+        zkUtils = ZkUtils.apply(
+            appConfig.getString(KafkaRestConfig.ZOOKEEPER_CONNECT_CONFIG), 30000,
+            30000,
+            JaasUtils.isZkSecurityEnabled()
+        );
+      }
+      if (mdObserver == null) {
+        mdObserver = new MetadataObserver(appConfig, zkUtils);
+      }
+      if (producerPool == null) {
+        producerPool = new ProducerPool(appConfig, zkUtils);
+      }
+      if (consumerManager == null) {
+        consumerManager = new ConsumerManager(appConfig, mdObserver);
+      }
+      if (simpleConsumerFactory == null) {
+        simpleConsumerFactory = new SimpleConsumerFactory(appConfig);
+      }
+      if (simpleConsumerManager == null) {
+        simpleConsumerManager =
+            new SimpleConsumerManager(appConfig, mdObserver, simpleConsumerFactory);
+      }
+      if (kafkaConsumerManager == null) {
+        kafkaConsumerManager = new KafkaConsumerManager(appConfig);
+      }
+
+      defaultZkUtils = zkUtils;
+      context =
+          new KafkaRestContextImpl(appConfig, mdObserver, producerPool, consumerManager,
+              simpleConsumerManager, kafkaConsumerManager
+          );
+      defaultAppConfig = appConfig;
+    }
+  }
+
+  public static KafkaRestContext getDefaultContext() {
```

Comment:
there is something sub-optimal with the factoring here i think - I don't think the concept of defaultContext is best exposed. maybe it would be better if there was some different type that encapsulated all these objects in addition to DefaultContext. It'd be pretty similar, just but different purpose. or separate methods, though that is more verbose. maybe what's here is not a bad compromise.